app/service/account.py

```diff
@@ -1,19 +1,30 @@
 """Account service."""
 
+from decimal import Decimal
 from uuid import UUID
 
+from app.constants import D0
 from app.repository.group import RepositoryGroup
 from app.schema.domain import ProjAccount, SysAccount, VlabAccount
+from app.service.budget import top_up
 
 
 async def add_system(repos: RepositoryGroup, account_id: UUID, name: str) -> SysAccount:
     """Add the system account."""
     return await repos.account.add_sys_account(account_id=account_id, name=name)
 
 
-async def add_virtual_lab(repos: RepositoryGroup, account_id: UUID, name: str) -> VlabAccount:
+async def add_virtual_lab(
+    repos: RepositoryGroup, account_id: UUID, name: str, balance: Decimal
+) -> VlabAccount:
     """Add a new virtual lab account."""
-    return await repos.account.add_vlab_account(account_id=account_id, name=name)
+    account = await repos.account.add_vlab_account(account_id=account_id, name=name)
+
+    if balance == D0:
+        return account
+
+    await top_up(repos, vlab_id=account.id, amount=balance)
+    return await repos.account.get_vlab_account(vlab_id=account.id)
```

Comment:
Looks good, but is the use case of creating an account with an initial balance frequent enough to justify the implementation of the logic in the service, instead of asking the client to make 2 calls (create and top-up)?

Reply:
It's a small optimisation measure and in VLab API depending on the user state it's 50/50: half of the requests will use non-zero initial balance, while the other half will not.